[bug] Selenium teardown timeout leaves stale test data and breaks CI retries

**Describe the bug**
The CI Test step retries against the same running stack/database, so a single transient selenium timeout
in `TestServices.tearDownClass` poisons every subsequent retry. The visible symptom is
`test_radius_user_registration` failing with `AssertionError: 400 != 201` and also `test_add_superuser`, but that's a cascade: attempt 1 created signup-user, its teardown timed out before deleting it, and because the test process
exited non-zero make stop never ran, so retries re-POST the same user against a dirty DB and the API
correctly returns 400.

**Steps To Reproduce**
1. Open a PR and let the Merge Tests workflow run the Test step (`openwisp/openwisp-utils/.github/actions/retry-command@master` running `make runtests && make stop, max_attempts: 5`).
2. Have any attempt pass the tests but hit a transient chromedriver timeout in tearDownClass →`_delete_object (cls.base_driver.get(...), ReadTimeoutError ... read timeout=120)`.
3. Observe signup-user is left in the DB and make stop is skipped (it's &&-chained after the failing command).
4. See attempts 2–5 fail with `AssertionError: 400 != 201` in `test_radius_user_registration` and `test_add_superuser` fail. Real-world occurrence: run 26470712885 (https://github.com/openwisp/docker-openwisp/actions/runs/26470712885).

**Expected behavior**
A transient selenium timeout in one attempt should not deterministically fail all retries. Each retry should start from clean state (or the test should be idempotent / clean up via the API), so the retry mechanism can actually recover.

**System Informatioon:**

- OS: GitHub Actions ubuntu-latest runner
- Docker version: GitHub Actions default (n/a CI-only)
- Browser and Browser Version: Headless Chormium

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -87,13 +87,21 @@ jobs:
           # /opt/openwisp/docker-openwisp. To ensure the test runs correctly
           # and environment variables remain intact, it is essential to
           # execute the test from this directory.
-          command: cd /opt/openwisp/docker-openwisp && make develop-pythontests && make stop
+          command: cd /opt/openwisp/docker-openwisp && make develop-pythontests
         env:
           SELENIUM_HEADLESS: 1
 
       - name: Print docker logs
         if: ${{ failure() }}
         run: docker compose logs
+
+      - name: Stop containers
+        # Run teardown exactly once after the retry loop, regardless of the
+        # test outcome. Keeping `make stop` out of the retried command ensures
+        # a failed attempt never skips teardown and never tears the stack down
+        # between retries (which would make every subsequent retry fail).
+        if: ${{ always() && steps.auto_install_upgrade.conclusion == 'success' }}
+        run: cd /opt/openwisp/docker-openwisp && make stop
 
       # the following action is equivalent to
       # echo "$DOCKER_HUB_SECRET" | docker login --username "$DOCKER_HUB_USERNAME" --password-stdin

--- a/tests/runtests.py
+++ b/tests/runtests.py
@@ -6,7 +6,7 @@ from urllib import error as urlerror
 from urllib import request
 
 import requests
-from selenium.common.exceptions import NoSuchElementException, TimeoutException
+from selenium.common.exceptions import TimeoutException
 from selenium.webdriver.common.by import By
 from selenium.webdriver.support import expected_conditions as EC
 from selenium.webdriver.support.ui import WebDriverWait
@@ -143,6 +143,43 @@ class TestServices(TestUtilities, unittest.TestCase):
         )
 
     @classmethod
+    def _cleanup_stale_test_data(cls):
+        """Delete test-created users that may remain from a previous failed run.
+
+        Runs before tests start so each CI attempt begins with a clean
+        slate. Uses the Django ORM directly (via docker compose exec) to
+        avoid any Selenium dependency during setup.
+        """
+        try:
+            cls._execute_docker_compose_command(
+                [
+                    "docker",
+                    "compose",
+                    "exec",
+                    "-T",
+                    "dashboard",
+                    "python",
+                    "manage.py",
+                    "shell",
+                    "-c",
+                    (
+                        "from openwisp_users.models import User; "
+                        "User.objects.filter("
+                        "username__in=['signup-user', 'test_superuser', "
+                        "'test_superuser2']"
+                        ").delete()"
+                    ),
+                ],
+                use_text_mode=True,
+            )
+        except Exception as e:
+            exc_type = type(e).__name__
+            print(
+                f"Warning: stale test data cleanup failed ({exc_type}: {e}). "
+                "Individual tests may fail if stale data is present."
+            )
+
+    @classmethod
     def setUpClass(cls):
         cls.failed_test = False
         cls.live_server_url = cls.config["app_url"]
@@ -171,6 +208,7 @@ class TestServices(TestUtilities, unittest.TestCase):
                 ["docker", "compose", "up", "--detach"],
             )
         cls._setup_admin_theme_links()
+        cls._cleanup_stale_test_data()
         # Create base drivers (Firefox)
         if cls.config["driver"] == "firefox":
             cls.base_driver = cls.get_firefox_webdriver()
@@ -186,8 +224,9 @@ class TestServices(TestUtilities, unittest.TestCase):
         for resource_link in cls.objects_to_delete:
             try:
                 cls._delete_object(resource_link)
-            except NoSuchElementException:
-                print(f"Unable to delete resource at: {resource_link}")
+            except Exception as e:
+                exc_type = type(e).__name__
+                print(f"Unable to delete resource at {resource_link}: {exc_type}: {e}")
         cls.second_driver.quit()
         cls.base_driver.quit()
         # Remove the temporary custom CSS file created for testing
@@ -419,7 +458,7 @@ class TestServices(TestUtilities, unittest.TestCase):
 
     def test_radius_user_registration(self):
         """Ensure users can register using the RADIUS API."""
-        url = f'{self.config["api_url"]}/api/v1/radius/organization/default/account/'
+        url = f"{self.config['api_url']}/api/v1/radius/organization/default/account/"
         response = requests.post(
             url,
             json={

